tools/build_langserver/langserver/analyzer.go

```diff
@@ -241,87 +269,171 @@ func (a *Analyzer) BuildLabelFromString(ctx context.Context, rootPath string,
-		labelPath = filepath
-	} else if strings.HasPrefix(label.PackageDir(), rootPath) {
-		labelPath = path.Join(label.PackageDir(), "BUILD")
-	} else {
-		labelPath = path.Join(rootPath, label.PackageDir(), "BUILD")
-	}
-
+	labelPath := string(a.BuildFileURIFromPackage(label.PackageDir()))
 	if !fs.PathExists(labelPath) {
 		return nil, fmt.Errorf("cannot find the path for build label %s", labelStr)
 	}
 
 	// Get the BuildDef and BuildDefContent for the BuildLabel
-	var buildDef *Identifier
-	var buildDefContent string
+	var buildDef *BuildDef
+	var definition string
 
-	// Check for cases such as "//tools/build_langserver/..."
 	if label.IsAllSubpackages() {
-		buildDefContent = "BuildLabel includes all subpackages in path: " +
-			path.Join(rootPath, label.PackageDir())
-
-		// Check for cases such as "//tools/build_langserver/all"
+		// Check for cases such as "//tools/build_langserver/..."
+		definition = "BuildLabel includes all subpackages in path: " + path.Join(path.Dir(labelPath))
 	} else if label.IsAllTargets() {
-		buildDefContent = "BuildLabel includes all BuildTargets in BUILD file: " + labelPath
+		// Check for cases such as "//tools/build_langserver/all"
+		definition = "BuildLabel includes all BuildTargets in BUILD file: " + labelPath
 	} else {
-		// Get the BuildDef IdentStatement from the build file
-		buildDef, err = a.getBuildDefByName(label.Name, labelPath)
-		if err != nil {
-			return nil, err
-		}
-
-		// Get the content for the BuildDef
-		labelfileContent, err := ReadFile(ctx, lsp.DocumentURI(labelPath))
+		buildDef, err = a.BuildDefFromLabel(ctx, &label, labelPath)
 		if err != nil {
 			return nil, err
 		}
-		buildDefContent = strings.Join(labelfileContent[buildDef.Pos.Line:buildDef.EndPos.Line+1], "\n")
+		definition = buildDef.Content
 	}
 
 	return &BuildLabel{
-		BuildLabel:      &label,
-		Path:            labelPath,
-		BuildDef:        buildDef,
-		BuildDefContent: buildDefContent,
+		BuildLabel: &label,
+		Path:       labelPath,
+		BuildDef:   buildDef,
+		Definition: definition,
 	}, nil
 }
 
-// getBuildDefByName returns an Identifier object of a BuildDef(call of a Build rule) based the name
-func (a *Analyzer) getBuildDefByName(name string, path string) (*Identifier, error) {
+// BuildDefFromLabel returns a BuildDef struct given an *core.BuildLabel and the path of the label
+func (a *Analyzer) BuildDefFromLabel(ctx context.Context, label *core.BuildLabel, path string) (*BuildDef, error) {
+	if label.IsAllSubpackages() || label.IsAllTargets() {
+		return nil, nil
+	}
+
+	// Get the BuildDef IdentStatement from the build file
+	buildDef, err := a.getBuildDefByName(ctx, label.Name, path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the content for the BuildDef
+	labelfileContent, err := ReadFile(ctx, lsp.DocumentURI(path))
+	if err != nil {
+		return nil, err
+	}
+	buildDef.Content = strings.Join(labelfileContent[buildDef.Pos.Line:buildDef.EndPos.Line+1], "\n")
```

Comment:
you sure about this? are the Pos's zero-indexed or 1-indexed?

Reply:
They are zero indexed! :) lsp.Positions! :)